### PR TITLE
#2694.Add more type inference tests

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

File: LanguageFeatures/Control-flow-collections/Null-aware-elements/type_inference_map_A05_t01.dart [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import '../../../Utils/static_type_helper.dart';
+
+main() {
+  String? ek = "ek";
+
+  var map = <String, int>{
+    ?(contextType(ek)..expectStaticType<Exactly<String?>>()): 1,
+    ?null: 2
+  };
+  map.expectStaticType<Exactly<Map<String, int>>>();
```
Comment (eernstg) [inline LanguageFeatures/Control-flow-collections/Null-aware-elements/type_inference_map_A05_t01.dart:63]: I don't think line 63 contributes to the goal of the test, and we have other tests showing that the inferred type of a local `var` variable is the static type of the initializing expression, and the static type of a map `<K, V>{...}` is `Map<K, V>`.

Or did I miss something which is actually new?

File: LanguageFeatures/Control-flow-collections/Null-aware-elements/type_inference_map_A06_t01.dart [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import '../../../Utils/static_type_helper.dart';
+
+main() {
+  String? ek = "ek";
+  int? ev = 1;
+
+  var map = <String, int>{
+    contextType(ek)..expectStaticType<Exactly<String>>()): ?ev,
+  };
+  map.expectStaticType<Exactly<Map<String, int>>>();
```
Comment (eernstg) [inline LanguageFeatures/Control-flow-collections/Null-aware-elements/type_inference_map_A06_t01.dart:63]: Ditto.

Comment (eernstg) [reply LanguageFeatures/Control-flow-collections/Null-aware-elements/type_inference_map_A06_t01.dart:63]: Same below.

File: LanguageFeatures/Control-flow-collections/Null-aware-elements/type_inference_map_A07_t01.dart [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+import '../../../Utils/static_type_helper.dart';
+
+main() {
+  String? ev = "ev";
+
+  var map = <String, String?>{
+    "ek": ?(contextType(ev)..expectStaticType<Exactly<String?>>())
+  };
+  map.expectStaticType<Exactly<Map<String, String>>>();
```
Comment (eernstg) [inline LanguageFeatures/Control-flow-collections/Null-aware-elements/type_inference_map_A07_t01.dart:62]: That would be `Map<String, String?>`, but I think this line can be deleted, like similar lines in earlier tests.

File: LanguageFeatures/Control-flow-collections/Null-aware-elements/type_inference_map_A08_t01.dart [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import '../../../Utils/static_type_helper.dart';
+
+main() {
+  String? ek = 2 > 1 ? "ek" : null;
+  String? ev = "ev";
+
+  var map = <String, String>{
+    ?ek: contextType(ev)..expectStaticType<Exactly<String>>(),
+  };
+  map.expectStaticType<Exactly<Map<String, String>>>();
```
Comment (eernstg) [inline LanguageFeatures/Control-flow-collections/Null-aware-elements/type_inference_map_A08_t01.dart:63]: Ditto.

**Review by eernstg — COMMENTED** [review]
Looks good!

One general thing came up: I think it's useful to adjust the descriptions such that they use standard terminology. I commented on what that would be in a comment on LanguageFeatures/Control-flow-collections/Null-aware-elements/type_inference_list_A01_t01.dart.

**Comment by sgrekhov** [comment]
Thank you! Descriptions updated, excessive checks removed. PTAL

File: LanguageFeatures/Control-flow-collections/Null-aware-elements/type_inference_map_A01_t01.dart [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+///         because the `?` will discard a `null` entry.
+///       - The inferred value element type is `NonNull(Uv)`. The entry added to
+///         the map will never have a `null` value.
+///     - Else the inferred value element type is the inferred type of `ev` in
+///       context `Pv`. The whole element is null-aware, but the value part is
+///       not, so it is inferred as normal.
+///
+/// @description Check that inferred type of element `?ek: ev` in an unknown
+/// context is `NonNull(Uk): ev`, where `Uk` is the inferred type of the
+/// expression `ek` in an unknown context.
```
[the remark this answers is not in the document]
Comment (eernstg) [reply LanguageFeatures/Control-flow-collections/Null-aware-elements/type_inference_map_A01_t01.dart:51]: Thanks!

**Review by eernstg — APPROVED** [review]
Great, thanks! LGTM

File: LanguageFeatures/Control-flow-collections/Null-aware-elements/type_inference_list_A01_t01.dart [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+///     - Let `U` be the inferred type of the expression `e1` in context `Ps?`.
+///       The expression has a nullable context type because it may safely
+///       evaluate to `null` even when the surrounding set doesn't allow that
+///       because the `?` will discard a `null` entry.
+///   - The inferred set element type is `NonNull(U)`. The value added to the
+///     set will never be `null`.
+///
+/// @description Check that the inferred type of element `?e1` in an unknown
+/// context is `NonNull(U)`, where `U` is the inferred type of the expression
+/// `e1` in an unknown context.
```
[the remark this answers is not in the document]
Comment (eernstg) [reply LanguageFeatures/Control-flow-collections/Null-aware-elements/type_inference_list_A01_t01.dart]: Thanks a lot for those adjustments!

**Pull request merged**